File: app/src/main/java/com/codesroots/osamaomar/shopgate/presentationn/screens/feature/home/subcategryfragment/SubcategryFragment.java
package com.codesroots.osamaomar.shopgate.presentationn.screens.feature.home.subcategryfragment;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.codesroots.osamaomar.shopgate.R;
import com.codesroots.osamaomar.shopgate.entities.SubCategriesWithProducts;
import com.codesroots.osamaomar.shopgate.helper.PreferenceHelper;
import com.codesroots.osamaomar.shopgate.presentationn.screens.feature.home.mainactivity.MainActivity;
import com.codesroots.osamaomar.shopgate.presentationn.screens.feature.home.subcategryfragment.adapters.MoreSalesProductsAdapter;
import com.codesroots.osamaomar.shopgate.presentationn.screens.feature.home.subcategryfragment.adapters.SubCatsAdapter;

import java.util.List;

import static com.codesroots.osamaomar.shopgate.entities.names.CAT_ID;
import static com.codesroots.osamaomar.shopgate.entities.names.CAT_NAME;

public class SubcategryFragment extends Fragment {

    private SubCatesViewModel mViewModel;
    RecyclerView subCates,MoreSaleProducts;
    private int catid,userid= PreferenceHelper.getUserId();
    private TextView product_notfound,text;
    private FrameLayout progress;
    private MoreSalesProductsAdapter moreSalesProductsAdapter;
    private  String categry_name;
    private List<SubCategriesWithProducts.ProductsbyrateBean> productsData;


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.subcategry_fragment, container, false);
        subCates = view.findViewById(R.id.subcategry_Rec);
        product_notfound = view.findViewById(R.id.product_notfound);
        MoreSaleProducts = view.findViewById(R.id.product_Rec);
        progress = view.findViewById(R.id.progress);
        text = view.findViewById(R.id.text);
        catid = getArguments().getInt(CAT_ID,0);
        categry_name = getArguments().getString(CAT_NAME);
   //     ((MainActivity)getActivity()).head_title.setText(categry_name);
        ((MainActivity)getActivity()).logo.setVisibility(View.VISIBLE);
        mViewModel = ViewModelProviders.of(this,getViewModelFactory()).get(SubCatesViewModel.class);

        mViewModel.subCategriesWithProductsMutableLiveData.observe(this,subCategriesWithProducts ->
                {
                    text.setVisibility(View.VISIBLE);
                    progress.setVisibility(View.GONE);
                    subCates.setAdapter(new SubCatsAdapter(getActivity(),subCategriesWithProducts.getData()));
                    if(subCategriesWithProducts.getProductsbyrate().size()>0) {
                        moreSalesProductsAdapter = new MoreSalesProductsAdapter(getActivity(), subCategriesWithProducts.getProductsbyrate(), mViewModel);
                        productsData = subCategriesWithProducts.getProductsbyrate();
                        MoreSaleProducts.setAdapter(moreSalesProductsAdapter);
                    }
                    else
                        product_notfound.setVisibility(View.VISIBLE);
                });

        mViewModel.throwableMutableLiveData.observe(this,throwable ->
                {
                    progress.setVisibility(View.GONE);
                    Snackbar.make(view,throwable.toString(),Snackbar.LENGTH_SHORT).show();
                });

        mViewModel.addToFavMutableLiveData.observe(this, addToFavModel -> {
            productsData.get(mViewModel.current_item).getProduct()
                    .getFavourites().add(new SubCategriesWithProducts.ProductsbyrateBean.ProductBean.FavouritesBean(addToFavModel.getLikeid()));
            moreSalesProductsAdapter.notifyItemChanged(mViewModel.current_item);
        });

        mViewModel.deleteToFavMutableLiveData.observe(this, aBoolean -> {
            productsData.get(mViewModel.current_item).getProduct()
                    .getFavourites().clear();
            moreSalesProductsAdapter.notifyDataSetChanged();
        });
        return  view;
    }

    private SubCatsViewModelFactory getViewModelFactory() {
        return new SubCatsViewModelFactory(this.getActivity().getApplication(),catid,userid
        );
    }

}

File: app/src/main/java/com/codesroots/osamaomar/cityrolls/presentationn/screens/feature/home/subcategryfragment/SubcategryFragment.kt
package com.codesroots.osamaomar.cityrolls.presentationn.screens.feature.home.subcategryfragment

import androidx.lifecycle.ViewModelProviders
import android.os.Bundle
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.TextView
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.LinearLayoutManager

import com.codesroots.osamaomar.cityrolls.R
import com.codesroots.osamaomar.cityrolls.entities.Category
import com.codesroots.osamaomar.cityrolls.entities.Items
import com.codesroots.osamaomar.cityrolls.helper.PreferenceHelper
import com.codesroots.osamaomar.cityrolls.presentationn.screens.feature.home.mainactivity.MainActivity

import com.codesroots.osamaomar.cityrolls.entities.names.CAT_ID
import com.codesroots.osamaomar.cityrolls.entities.names.SUBCATES_NAME
import com.codesroots.osamaomar.cityrolls.presentationn.screens.feature.home.subcategryfragment.MainViewModelFactory
import com.codesroots.osamaomar.cityrolls.presentationn.screens.feature.home.mainfragment.adapters.CatAdapter
import com.codesroots.osamaomar.cityrolls.presentationn.screens.feature.home.subcategryfragment.adapters.SubCatsAdapter
import kotlinx.android.synthetic.main.subcategry_fragment.view.*

class SubcategryFragment : Fragment() {

    private var viewModel: MainFragmentViewModel? = null
    private var catid: Int = 0
    private val userid = PreferenceHelper.getUserId()
    private var product_notfound: TextView? = null
    private var text: TextView? = null
    private var progress: FrameLayout? = null
    private var categry_name: String? = null
    var adapterr: SubCatsAdapter? = null
    var data: Category? = null
    var datArray : List<Items>? = null
    private fun getViewModelFactory(): MainViewModelFactory {
        return MainViewModelFactory(this.activity!!.application)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.subcategry_fragment, container, false)
        product_notfound = view.findViewById(R.id.product_notfound)
        progress = view.findViewById(R.id.progress)
        text = view.findViewById(R.id.text)

        catid = arguments!!.getInt(CAT_ID, 0)
             data = arguments!!.getParcelable("data")

      //  categry_name = arguments!!.getString(SUBCATES_NAME)

        if (data != null) {
        datArray = data!!.items
        adapterr!!.notifyDataSetChanged()


        }

        (activity as MainActivity).head_title.text = categry_name
        (activity as MainActivity).logo.visibility = View.VISIBLE

        viewModel = ViewModelProviders.of(this, getViewModelFactory()).get(MainFragmentViewModel::class.java)
        viewModel!!.mainViewMutableLiveData?.observe(this, androidx.lifecycle.Observer {
            datArray = it.items.get(0).items
            adapterr = SubCatsAdapter(context as FragmentActivity, datArray!!)
            view.subcategry_Rec.layoutManager = LinearLayoutManager(activity, RecyclerView.HORIZONTAL, false)
            view.subcategry_Rec?.adapter = adapterr

        })

        return view
    }

}
